packages/store/addon/-private/system/store/internal-model-factory.ts

```diff
@@ -0,0 +1,228 @@
+import coerceId from '../coerce-id';
+import { assert, warn } from '@ember/debug';
+import InternalModel from '../model/internal-model';
+import Store from '../store';
+import IdentityMap from '../identity-map';
+import InternalModelMap from '../internal-model-map';
+import { isNone } from '@ember/utils';
+import { JsonApiResourceIdentity } from '../../ts-interfaces/record-data-json-api';
+
+const FactoryCache = new WeakMap<object, InternalModelFactory>();
+
+let globalClientIdCounter = 1;
+
+export function internalModelFactoryFor(store: Store): InternalModelFactory {
+  let factory = FactoryCache.get(store);
+
+  if (factory === undefined) {
+    factory = new InternalModelFactory(store);
+    FactoryCache.set(store, factory);
+  }
+
+  return factory;
+}
+
+/**
+ * The InternalModelFactory handles the lifecyle of
+ * instantiating, caching, and destroying InternalModel
+ * instances.
+ *
+ * @internal
+ */
+export default class InternalModelFactory {
```

Comment:
Fwiw, it seems a bit odd to me that something named a factory would also keep a cache of things.